framework/build.gradle

```diff
@@ -85,20 +85,93 @@ checkCompilerMessages {
+    description 'Internal task to run the whole-program-inference tests with -Ainfer=stubs to generate stub files'
+
+    doFirst {
+        delete("tests/whole-program-inference/annotated")
+    }
+    outputs.upToDateWhen { false }
+    include '**/WholeProgramInferenceStubsTest.class'
+    testLogging {
+        // Always run the tests
+        outputs.upToDateWhen { false }
+
+        // Show the found unexpected diagnostics and the expected diagnostics not found.
+        exceptionFormat "full"
+        events "passed", "skipped", "failed"
+    }
+}
+
+task testWPIStubsValidate(type: Test) {
+    description 'Internal task to re-run the whole-program-inference tests using the stub files generated by testWPIStubs'
+
+    outputs.upToDateWhen { false }
+    include '**/WholeProgramInferenceStubsValidationTest.class'
+    testLogging {
+        // Always run the tests
+        outputs.upToDateWhen { false }
+
+        // Show the found unexpected diagnostics and the expected diagnostics not found.
+        exceptionFormat "full"
+        events "passed", "skipped", "failed"
+    }
+}
+
+// Common logic for copying and deleting whole-program inference tests.
+void commonWPIlogic() {
+    // Copying all test files to another directory, removing all expected errors that should not
+    // occur after inserting inferred annotations from .jaif files.
+    copy {
+        from files('tests/whole-program-inference/non-annotated')
+        into file('tests/whole-program-inference/annotated')
+        filter { String line ->
+            line.contains('// :: error:') ? null : line
+        }
+    }
+    // The only file for which expected errors are maintained is ExpectedErrors.java, so we copy it over.
+    delete('tests/whole-program-inference/annotated/ExpectedErrors.java')
+    copy {
+        from file('tests/whole-program-inference/non-annotated/ExpectedErrors.java')
+        into file('tests/whole-program-inference/annotated')
+    }
+}
+
+// This task is similar to the wholeProgramInferenceTests task below, but it doesn't
+// run the insert-annotations-to-source tool. Instead, it tests the -Ainfer=stubs feature
+// and the -AmergeStubsWithSource feature to do WPI using stub files.
+task wholeProgramInferenceStubTests(dependsOn: 'shadowJar', group: 'Verification') {
+    description 'Run tests for whole-program inference using stub files'
     dependsOn(compileTestJava)
+    dependsOn(testWPIStubs)
+    outputs.upToDateWhen { false }
+
+    doLast {
+        commonWPIlogic()
+        // The stub file format doesn't support annotations on anonymous inner classes, so
+        // this test also expects errors on UsesAnonymous.java.
+        delete('tests/whole-program-inference/annotated/UsesAnonymous.java')
+        copy {
+            from file('tests/whole-program-inference/non-annotated/UsesAnonymous.java')
+            into file('tests/whole-program-inference/annotated')
+        }
+    }
+    finalizedBy(testWPIStubsValidate)
```

Comment:
@smillst @kelloggm
Why is this expressed via a `finalizedBy` relationship rather than a `dependsOn` relationship?

The information at
  https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:finalizer_tasks
  https://discuss.gradle.org/t/when-should-one-use-dependson-and-when-finalizedby/4066
says that `finalizedBy` is primarily used for tasks that do cleanup, and `dependsOn` is for tasks that consume another task's output, which seems to be the case here.

I ask because running
```
./gradlew clean :framework:testWPIStubsValidate :framework:wholeProgramInferenceStubTests
```

fails.  Other gradle invocations also sometimes fail.  Since these failures seem to be nondeterministic (maybe they depend on the order in which tasks are executed), they are very hard to reproduce and diagnose.

There is one other suspicious use of `finalizedBy` in this file, which was added in 0822b40cb279bb858a7374e69485001868ffd594.

Reply:
(Moving the discussion back into its thread.)
@kelloggm Thanks for the change, but it doesn't seem to help.
Running
```
./gradlew clean :framework:testWPIStubsValidate :framework:wholeProgramInferenceStubTests
```
still fails for me.  The failure I am getting is:
```
tests.wpirunners.WholeProgramInferenceStubsValidationTest > run[whole-program-inference/annotated] FAILED
    java.lang.AssertionError: 29 out of 29 expected diagnostics were found.
    2 unexpected diagnostics were found:
    OverriddenMethodsTest.java:22: error: (override.param.invalid)
    ConflictingAnnotationsTest.java:8: error: (return.type.incompatible)
```

Now, running just
```
./gradlew clean :framework:wholeProgramInferenceStubTests
```
or
```
./gradlew clean :framework:testWPIStubsValidate
```
fails in the same way.  (That wasn't the case before, I *think*.)

Reply:
@mernst I still cannot reproduce these failures locally. I tried pulling from `typetools/stubparser`'s master branch, but there were no new commits there.

Can you copy the contents of `framework/build/whole-program-inference/OverriddenMethodsTestChild.astub`, `framework/build/whole-program-inference/OverriddenMethodsTestParent.astub`, and `framework/build/whole-program-inference/ConflictingAnnotationsTest.astub` here? I've pasted my copies of those below; if there are any discrepancies between them maybe that will give us a clue about what might be happening.

(unrelated, I think I just discovered `ConflictingAnnotationsTest` isn't testing what I thought it was. I'll try to fix that, but it shouldn't matter for this issue)

```
> cat framework/build/whole-program-inference/OverriddenMethodsTestChild.astub 
import testlib.wholeprograminference.qual.DefaultType;
import testlib.wholeprograminference.qual.Sibling1;
import testlib.wholeprograminference.qual.Sibling2;

class OverriddenMethodsTestChild {

  // methods:

  void foo(@DefaultType OverriddenMethodsTestChild this, @Sibling1 java.lang.Object obj, @Sibling2 java.lang.Object obj2);

  void bar(@Sibling1 OverriddenMethodsTestChild this, @Sibling2 java.lang.Object obj);

}
```

```
> cat framework/build/whole-program-inference/OverriddenMethodsTestParent.astub 
import testlib.wholeprograminference.qual.Sibling1;
import testlib.wholeprograminference.qual.Sibling2;

class OverriddenMethodsTestParent {

  // methods:

  void thud(@Sibling1 java.lang.Object obj1, @Sibling2 java.lang.Object obj2);

  void qux(@Sibling1 java.lang.Object obj1, @Sibling2 java.lang.Object obj2);

  OverriddenMethodsTestParent();

  void foo(@testlib.wholeprograminference.qual.Sibling1 java.lang.Object obj, @testlib.wholeprograminference.qual.Sibling2 java.lang.Object obj2);

}
```

```
> cat framework/build/whole-program-inference/ConflictingAnnotationsTest.astub 
import testlib.wholeprograminference.qual.Sibling1;
import testlib.wholeprograminference.qual.WholeProgramInferenceBottom;

class ConflictingAnnotationsTest {

  // methods:

  @WholeProgramInferenceBottom @Sibling1 int getLocalSibling1();

  @Sibling1 int getWPINamespaceSibling1();

}
```

Reply:
The contents are different for me.
```
> cat framework/build/whole-program-inference/OverriddenMethodsTestChild.astub
import testlib.wholeprograminference.qual.DefaultType;
import testlib.wholeprograminference.qual.Sibling1;
import testlib.wholeprograminference.qual.Sibling2;

class OverriddenMethodsTestChild {

  // methods:

  void foo(@DefaultType OverriddenMethodsTestChild this, @Sibling1 java.lang.Object obj, @Sibling2 java.lang.Object obj2);

  void bar(@Sibling1 OverriddenMethodsTestChild this, @Sibling2 java.lang.Object obj);

}
```
```
> cat framework/build/whole-program-inference/OverriddenMethodsTestParent.astub 
import testlib.wholeprograminference.qual.Sibling1;
import testlib.wholeprograminference.qual.Sibling2;

class OverriddenMethodsTestParent {

  // methods:

  void thud(@Sibling1 java.lang.Object obj1, @Sibling2 java.lang.Object obj2);

  void qux(@Sibling1 java.lang.Object obj1, @Sibling2 java.lang.Object obj2);

  OverriddenMethodsTestParent();

  void foo(@testlib.wholeprograminference.qual.Sibling1 java.lang.Object obj, @testlib.wholeprograminference.qual.Sibling2 java.lang.Object obj2);

}
```
```
> cat framework/build/whole-program-inference/ConflictingAnnotationsTest.astub 
import testlib.wholeprograminference.qual.Sibling1;
import testlib.wholeprograminference.qual.DefaultType;

class ConflictingAnnotationsTest {

  // methods:

  @DefaultType java.lang.Object getLocalSibling1(java.lang.Object o);

  void expectsSibling1(@testlib.wholeprograminference.qual.Sibling1 int i);

  @testlib.wholeprograminference.qual.Sibling1 int getSibling1();

  @Sibling1 int getWPINamespaceSibling1();

}
```